src/main.rs

```diff
@@ -26,30 +28,56 @@ fn get_default_device() -> Option<&'static str> {
         .map(|&str| str)
 }
 
+#[derive(Debug, Clone)]
 struct ProgramState {
     controllers: HashSet<controller::Controller>,
-    nodes: HashSet<Node>,
-    nodes_map: HashMap<controller::Controller, Vec<Node>>,
+    nodes: BTreeSet<Node>,
+    nodes_map: HashMap<controller::Controller, BTreeSet<Node>>,
+    value_ids: BTreeSet<ValueID>,
 }
 
 impl ProgramState {
     fn new() -> ProgramState {
         ProgramState {
             controllers: HashSet::new(),
-            nodes: HashSet::new(),
-            nodes_map: HashMap::new()
+            nodes: BTreeSet::new(),
+            nodes_map: HashMap::new(),
+            value_ids: BTreeSet::new()
         }
     }
+
+    pub fn add_node(&mut self, controller: controller::Controller, node: Node) {
+        let node_set = self.nodes_map.entry(controller).or_insert(BTreeSet::new());
+        node_set.insert(node);
+        self.nodes.insert(node);
+    }
+
+    pub fn remove_node(&mut self, controller: controller::Controller, node: Node) {
+        if let Some(node_set) = self.nodes_map.get_mut(&controller) {
+            node_set.remove(&node);
+        }
+        self.nodes.remove(&node);
+    }
```

Comment:
thinking in terms of API, what do you think of adding a `get_controller` method to the `Node` class, since we have home_id ? And then here we wouldn'd need the `controller` parameters.